connect-inject/handler.go

```diff
@@ -291,26 +207,39 @@ func (h *Handler) Handle(_ context.Context, req admission.Request) admission.Res
 	pod.Spec.InitContainers = append(pod.Spec.InitContainers, initContainer)
 
 	// Add the Envoy and Consul sidecars.
-	esContainer, err := h.envoySidecar(pod, req.Namespace)
+	envoySidecar, err := h.envoySidecar(pod, req.Namespace)
 	if err != nil {
 		h.Log.Error("Error configuring injection sidecar container", "err", err, "Request Name", req.Name)
 		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("error configuring injection sidecar container: %s", err))
 	}
-	pod.Spec.Containers = append(pod.Spec.Containers, esContainer)
+	pod.Spec.Containers = append(pod.Spec.Containers, envoySidecar)
 
-	connectContainer, err := h.consulSidecar(pod)
+	// Now that the consul-sidecar no longer needs to re-register services periodically
+	// (that functionality lives in the endpoints-controller),
+	// we only need the consul sidecar to run the metrics merging server.
+	// First, determine if we need to run the metrics merging server.
+	shouldRunMetricsMerging, err := h.shouldRunMergedMetricsServer(pod)
 	if err != nil {
-		h.Log.Error("Error configuring consul sidecar container", "err", err, "Request Name", req.Name)
-		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("error configuring consul sidecar container: %s", err))
+		h.Log.Error("Error determining if metrics merging server should be run", "err", err, "Request Name", req.Name)
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("error determining if metrics merging server should be run: %s", err))
+	}
+
+	// Add the consul-sidecar only if we need to run the metrics merging server.
+	if shouldRunMetricsMerging {
+		consulSidecar, err := h.consulSidecar(pod)
+		if err != nil {
+			h.Log.Error("Error configuring consul sidecar container", "err", err, "Request Name", req.Name)
+			return admission.Errored(http.StatusInternalServerError, fmt.Errorf("error configuring consul sidecar container: %s", err))
+		}
+		pod.Spec.Containers = append(pod.Spec.Containers, consulSidecar)
 	}
-	pod.Spec.Containers = append(pod.Spec.Containers, connectContainer)
 
 	// pod.Annotations has already been initialized by h.defaultAnnotations()
 	// and does not need to be checked for being a nil value.
 	pod.Annotations[annotationStatus] = injected
 
 	// Add annotations for metrics.
-	err = h.prometheusAnnotations(&pod)
+	err = h.prometheusAnnotations(pod)
```

Comment:
Also, given we are actually updating the pod's annotations for metrics in `prometheusAnnotations` I *think*, we should send a pointer reference and not a value one.

Reply:
Oh you're right! I totally missed that, good catch!

Reply:
Even though the tests didn't fail because `pod.Annotations` is a pointer/map, I've changed it to a pointer to make it more obvious that this function mutates an object.